### PR TITLE
Avoid type inference failure due to a typing template rejection

### PR DESCRIPTION
**Pull request by sklam**
Closes #2549.

This PR offers an alternative to #2549 and provides a general fix for the underlying problem.  

**Background**

A function can have multiple typing templates (the "type declarations") for different definitions due to overloading.  Currently, these templates can reject a use by returning None or by raising an exception (usually TypeError or TypingError).  The former is from legacy old.  The latter provides better user feedback by including the reason for the rejection.  However, the latter will also cause the function resolution to fail and eventually failing the type inference of the entire function.  

**The Fix**

This patch adds error capturing logic to the function resolution so that failure from earlier templates will not fail the resolution immediately.  Instead, the error information is captured.  An error is raised only if all templates reject the use site.  In such case, the error message looks like:

```text
Invalid usage of Function(<class 'float'>) with parameters (complex128)
 * parameterized
In definition 0:
    TypeError: float() does not support complex
    raised from /Users/siu/dev/numba/numba/typing/builtins.py:794
In definition 1:
    rejected
```

The failing reason for each template (named "definition" in the error) is enumerated.  The "rejected" message occurs when the template simply returned None.

**Further improvements**

I noticed several opportunities for refactoring and improvements but have decided to defer that work to the future as they involve complicated changes.  They are:

1. Enforces that templates must raise to reject, such that it is illegal to just return None.  This will provide better diagnostic on resolution failure.
2. The operators are currently handled differently.  The type inferrer uses the template directly instead of using the `numba.types.Function.get_call_types`, where the new error capturing logic is inserted.  This direct use of template makes overloading operators different from overloading other functions.  Also, it is causing some duplication of resolution logic and error handling logic.
3. Currently, the resolution stops at the first matching template.  However, we should scan for all templates and see if there are conflicts.

### CODE REVIEW

File: numba/types/functions.py [hunk 1]
```diff
@@ -43,23 +105,30 @@ def get_impl_key(self, sig):
...
-            if sig is not None:
-                self._impl_keys[sig.args] = temp.get_impl_key(sig)
-                return sig
+                if sig is not None:
+                    self._impl_keys[sig.args] = temp.get_impl_key(sig)
+                    return sig
+                else:
+                    failures.add_error(temp_cls, "rejected")
+
+        if failures:
```
Comment (ehsantn) [inline numba/types/functions.py]: Is this `if` needed @sklam ?

Comment (sklam) [reply numba/types/functions.py]: You're right.  I will turn it into a check for the unexpected case `bool(failures)==False`

File: numba/types/functions.py [hunk 2]
```diff
@@ -43,23 +105,33 @@ def get_impl_key(self, sig):
...
-            if sig is not None:
-                self._impl_keys[sig.args] = temp.get_impl_key(sig)
-                return sig
+                if sig is not None:
+                    self._impl_keys[sig.args] = temp.get_impl_key(sig)
+                    return sig
+                else:
+                    failures.add_error(temp_cls, "rejected")
+
+        if not failures:
```
Comment (ehsantn) [inline numba/types/functions.py]: `not failures` is always True here. I think you meant to check the length of the list of errors inside it: `if not failures._failures:`.

Comment (sklam) [reply numba/types/functions.py]: oops, must be from my leftover memory when `failures` was a `list`.  I'm glad you noticed that.

**Pull request merged**